fieldtracing/fieldtracing.cpp

```diff
@@ -996,114 +1010,213 @@ namespace FieldTracing {
-                  MPI_Allreduce(cellFWTracingStepSize.data(), reducedCellFWTracingStepSize.data(), globalDccrgSize, MPI_FLOAT, MPI_MAX, MPI_COMM_WORLD);
-                  MPI_Allreduce(cellBWTracingStepSize.data(), reducedCellBWTracingStepSize.data(), globalDccrgSize, MPI_FLOAT, MPI_MAX, MPI_COMM_WORLD);
+                  MPI_Allreduce(smallCellFWTracingCoordinates.data(), smallSumCellFWTracingCoordinates.data(), 3*smallSizeFW, MPI_FLOAT, MPI_SUM, MPI_COMM_WORLD);
+                  MPI_Allreduce(smallCellBWTracingCoordinates.data(), smallSumCellBWTracingCoordinates.data(), 3*smallSizeBW, MPI_FLOAT, MPI_SUM, MPI_COMM_WORLD);
+                  MPI_Allreduce(smallCellFWTracingStepSize.data(), smallReducedCellFWTracingStepSize.data(), smallSizeFW, MPI_FLOAT, MPI_MAX, MPI_COMM_WORLD);
+                  MPI_Allreduce(smallCellBWTracingStepSize.data(), smallReducedCellBWTracingStepSize.data(), smallSizeBW, MPI_FLOAT, MPI_MAX, MPI_COMM_WORLD);
+                  MPI_Allreduce(smallCellFWRunningDistance.data(), smallReducedCellFWRunningDistance.data(), smallSizeFW, MPI_FLOAT, MPI_MAX, MPI_COMM_WORLD);
+                  MPI_Allreduce(smallCellBWRunningDistance.data(), smallReducedCellBWRunningDistance.data(), smallSizeBW, MPI_FLOAT, MPI_MAX, MPI_COMM_WORLD);
+               }
+               for(int n=0; n<smallSizeFW; n++) {
+                  reducedCellFWTracingStepSize[indicesToReduceFW[n]] = smallReducedCellFWTracingStepSize[n];
+                  reducedCellFWRunningDistance[indicesToReduceFW[n]] = smallReducedCellFWRunningDistance[n];
+                  reducedCellNeedsContinuedFWTracing[indicesToReduceFW[n]] = smallReducedCellNeedsContinuedFWTracing[n];
+                  reducedCellFWConnection[indicesToReduceFW[n]] = smallReducedCellFWConnection[n];
+                  sumCellFWTracingCoordinates[indicesToReduceFW[n]] = smallSumCellFWTracingCoordinates[n];
+               }
+               for(int n=0; n<smallSizeBW; n++) {
+                  reducedCellBWTracingStepSize[indicesToReduceBW[n]] = smallReducedCellBWTracingStepSize[n];
+                  reducedCellBWRunningDistance[indicesToReduceBW[n]] = smallReducedCellBWRunningDistance[n];
+                  reducedCellNeedsContinuedBWTracing[indicesToReduceBW[n]] = smallReducedCellNeedsContinuedBWTracing[n];
+                  reducedCellBWConnection[indicesToReduceBW[n]] = smallReducedCellBWConnection[n];
+                  sumCellBWTracingCoordinates[indicesToReduceBW[n]] = smallSumCellBWTracingCoordinates[n];
                }
-               anyCellNeedsTracing = false;
             }
             #pragma omp barrier
             phiprof::stop("MPI-loop");
             #pragma omp single
             {
-               cellsToDo = 0;
+               cellsToDoFullBox = 0;
+               cellsToDoFluxRopes = 0;
+               cellFWTracingStepSize = reducedCellFWTracingStepSize;
```

Comment:
These are vector deep copies. You could do a swap() on them instead to reduce data flow.

A deep copy is needed for the NeedsContinuedXWTracing vectors though, as you apparently use the "old" one to keep track of active tracings still.

Reply:
These are still deep copies?